File: clients/windows/Mailcal.Tests/FolderActionsTests.cs
// Changing the folder tree from the pane (docs/folder-pane.md, "Changing the tree", rules 22 to 28):
// what each row's menu carries, where Move to… may send a folder, which drops land, and the words
// the dialogs and the notice use.
//
// Every one of these fails silently on screen. A folder offered a Delete it may not have is a
// click the core ignores; a Move to… that lists the folder's own subfolder moves it nowhere; a drop
// taken across accounts moves nothing and looks as if it did. None of it is reachable from a
// screenshot, and the WinUI half cannot be linked into this assembly.

using System.Collections.ObjectModel;
using Allodia.Mailcal.Services;
using Allodia.Mailcal.ViewModels;
using uniffi.mailcal_bindings;
using Xunit;

namespace Allodia.Mailcal.Tests;

public class FolderActionsTests
{
    private const string Account = "acct-1";

    private static readonly FolderNameWords NameWords =
        new("empty", "surrounded", "control", "separator", "taken");

    private static FolderItem Folder(
        string key,
        string name,
        string? parent = null,
        bool editable = true,
        bool acceptsFolders = true,
        bool pending = false,
        SidebarFolderRole role = SidebarFolderRole.None) =>
        new()
        {
            Key = key,
            Name = name,
            Parent = parent,
            Role = role,
            Editable = editable,
            AcceptsFolders = acceptsFolders,
            AcceptsMessages = !pending,
            Pending = pending,
        };

    /// <summary>Inbox, then `work` holding `clients` holding `acme`, then `bills` at the top.</summary>
    private static List<FolderItem> Tree() =>
        [
            Folder("inbox", "Inbox", editable: false, role: SidebarFolderRole.Inbox),
            Folder("work", "Work"),
            Folder("clients", "Clients", parent: "work"),
            Folder("acme", "Acme", parent: "clients"),
            Folder("bills", "Bills"),
        ];

    private static SidebarItem FolderRow(bool editable, bool acceptsFolders, bool pending = false) =>
        new()
        {
            Tag = "work",
            OwnerAccountId = Account,
            Editable = editable,
            AcceptsFolders = acceptsFolders,
            IsPending = pending,
        };

    [Fact]
    public void AFolderRowOffersExactlyWhatItsFlagsAllow()
    {
        Assert.Equal(
            new[] { FolderMenuAction.NewFolder, FolderMenuAction.Rename, FolderMenuAction.MoveTo, FolderMenuAction.Delete },
            FolderActions.Menu(FolderRow(editable: true, acceptsFolders: true)));
        // A role folder (the Inbox) takes subfolders and is named and placed by the app.
        Assert.Equal(
            new[] { FolderMenuAction.NewFolder },
            FolderActions.Menu(FolderRow(editable: false, acceptsFolders: true)));
        // A folder inside Trash can be restored or removed, and takes no new subfolders.
        Assert.Equal(
            new[] { FolderMenuAction.Rename, FolderMenuAction.MoveTo, FolderMenuAction.Delete },
            FolderActions.Menu(FolderRow(editable: true, acceptsFolders: false)));
        // Junk, Drafts: nothing, so no menu at all.
        Assert.Empty(FolderActions.Menu(FolderRow(editable: false, acceptsFolders: false)));
    }

    [Fact]
    public void APendingFolderOffersNothingWhateverItsFlagsSay()
    {
        // Its key may still move on the server, so nothing may be built on it (rule 27).
        Assert.Empty(FolderActions.Menu(FolderRow(editable: true, acceptsFolders: true, pending: true)));
    }

    [Fact]
    public void AnAccountRowOffersNewFolderOnlyWhereItsFoldersCanChange()
    {
        var managed = new SidebarItem { Tag = "acct:a", AccountId = "a", AcceptsFolders = true };
        var fixedTree = new SidebarItem { Tag = "acct:b", AccountId = "b", AcceptsFolders = false };
        Assert.Equal(
            new[] { FolderMenuAction.NewFolder, FolderMenuAction.RemoveAccount },
            FolderActions.Menu(managed));
        Assert.Equal(new[] { FolderMenuAction.RemoveAccount }, FolderActions.Menu(fixedTree));
    }

    [Fact]
    public void TheGroupAndTheSyntheticRowsCarryNoMenu()
    {
        Assert.Empty(FolderActions.Menu(new SidebarItem { Tag = SidebarTree.AllAccountsTag, IsGroup = true }));
        Assert.Empty(FolderActions.Menu(new SidebarItem { Tag = SidebarTree.AddAccountTag }));
    }

    [Fact]
    public void MoveToListsTheTopLevelFirstAndNeverTheFolderOrWhatIsInsideIt()
    {
        var targets = FolderActions.MoveTargets(Tree(), "clients", "Top level");

        Assert.Equal((string?)null, targets[0].Key);
        Assert.Equal("Top level", targets[0].Label);
        var keys = targets.Skip(1).Select(t => t.Key).ToList();
        Assert.DoesNotContain("clients", keys);
        Assert.DoesNotContain("acme", keys);
        Assert.Contains("work", keys);
        Assert.Contains("bills", keys);
    }

    [Fact]
    public void MoveToNamesEachFolderWithTheFoldersItSitsIn()
    {
        var targets = FolderActions.MoveTargets(Tree(), "bills", "Top level");
        Assert.Contains(targets, t => t.Key == "acme" && t.Label == "Work / Clients / Acme");
    }

    [Fact]
    public void MoveToLeavesOutAFolderThatTakesNoSubfolders()
    {
        var tree = Tree();
        tree.Add(Folder("trash", "Trash", editable: false, acceptsFolders: false, role: SidebarFolderRole.Trash));
        Assert.DoesNotContain(FolderActions.MoveTargets(tree, "bills", "Top level"), t => t.Key == "trash");
    }

    [Fact]
    public void AFolderDropLandsOnlyWhereItCanGoWithinItsAccount()
    {
        var tree = Tree();
        var drag = new FolderDrag(Account, "clients");

        Assert.True(FolderActions.TakesFolder(tree, drag, Account, "bills", targetAcceptsFolders: true));
        Assert.True(FolderActions.TakesFolder(tree, drag, Account, null, targetAcceptsFolders: true),
            "its account's row moves it to the top");
        Assert.False(FolderActions.TakesFolder(tree, drag, "acct-2", "bills", targetAcceptsFolders: true),
            "a folder never crosses accounts");
        Assert.False(FolderActions.TakesFolder(tree, drag, Account, "acme", targetAcceptsFolders: true),
            "a folder cannot move inside itself");
        Assert.False(FolderActions.TakesFolder(tree, drag, Account, "clients", targetAcceptsFolders: true));
        Assert.False(FolderActions.TakesFolder(tree, drag, Account, "work", targetAcceptsFolders: true),
            "dropped back where it already is, it does not move");
        Assert.False(FolderActions.TakesFolder(tree, drag, Account, "bills", targetAcceptsFolders: false));
    }

    [Fact]
    public void ATopLevelFolderDroppedOnItsOwnAccountIsNotAMove()
    {
        Assert.False(FolderActions.TakesFolder(
            Tree(), new FolderDrag(Account, "bills"), Account, null, targetAcceptsFolders: true));
    }

    [Fact]
    public void MessagesLandOnlyOnAFolderOfTheirOwnAccountThatTakesMail()
    {
        var mine = new MessageDrag(
            [new SelectedRow.Message(Account, "m1"), new SelectedRow.Thread(Account, "t1")]);
        var mixed = new MessageDrag(
            [new SelectedRow.Message(Account, "m1"), new SelectedRow.Message("acct-2", "m2")]);

        Assert.True(FolderActions.TakesMessages(mine, Account, targetAcceptsMessages: true));
        Assert.False(FolderActions.TakesMessages(mine, Account, targetAcceptsMessages: false),
            "Junk takes mail through a report, never a drop");
        Assert.False(FolderActions.TakesMessages(mine, "acct-2", targetAcceptsMessages: true));
        Assert.False(FolderActions.TakesMessages(mixed, Account, targetAcceptsMessages: true),
            "a selection spanning accounts would move only part of itself");
        Assert.False(FolderActions.TakesMessages(new MessageDrag([]), Account, targetAcceptsMessages: true));
    }

    [Fact]
    public void ATreeThatLoopsDoesNotHangTheWalk()
    {
        List<FolderItem> looped = [Folder("a", "A", parent: "b"), Folder("b", "B", parent: "a")];
        Assert.False(FolderActions.IsInside(looped, "a", "c"));
        Assert.True(FolderActions.IsInside(looped, "a", "b"));
    }

    [Fact]
    public void TheNameDialogConfirmsOnlyAValidNameAndSaysWhyOtherwise()
    {
        Assert.Equal((true, string.Empty), FolderActions.NameState(FolderNameCheck.Valid, NameWords));
        Assert.Equal((false, "empty"), FolderActions.NameState(FolderNameCheck.Empty, NameWords));
        Assert.Equal((false, "surrounded"), FolderActions.NameState(FolderNameCheck.Surrounded, NameWords));
        Assert.Equal((false, "control"), FolderActions.NameState(FolderNameCheck.Control, NameWords));
        Assert.Equal((false, "separator"), FolderActions.NameState(FolderNameCheck.Separator, NameWords));
        Assert.Equal((false, "taken"), FolderActions.NameState(FolderNameCheck.Taken, NameWords));
    }

    [Fact]
    public void TheDeleteQuestionSaysTrashOutsideTrashAndForGoodInsideIt()
    {
        var words = new DeleteWords(
            name => $"Delete {name}?", "into Trash", "Move to Trash",
            name => $"Delete {name} permanently?", "for good", "Delete permanently");

        var outside = FolderActions.Delete(inTrash: false, "Bills", words);
        var inside = FolderActions.Delete(inTrash: true, "Bills", words);

        Assert.Equal(new DeleteQuestion("Delete Bills?", "into Trash", "Move to Trash"), outside);
        Assert.Equal(new DeleteQuestion("Delete Bills permanently?", "for good", "Delete permanently"), inside);
    }

    [Fact]
    public void TheNoticeNamesTheFolderAndWhyTheChangeDidNotHappen()
    {
        static string Text(FolderNotice? notice) =>
            FolderActions.NoticeText(notice, name => $"changed:{name}", name => $"refused:{name}");

        Assert.Equal(string.Empty, Text(null));
        Assert.Equal("changed:Work", Text(new FolderNotice(Account, "Work", FolderAction.Rename, FolderProblem.ChangedElsewhere)));
        Assert.Equal("refused:Work", Text(new FolderNotice(Account, "Work", FolderAction.Delete, FolderProblem.Refused)));
    }

    [Fact]
    public void TheReconcileCopiesWhatEachRowOffersAndWhetherItWaits()
    {
        var target = new ObservableCollection<SidebarItem>();
        var account = new AccountItem
        {
            Id = Account,
            Email = "a@example.com",
            SendLabel = "a@example.com",
            Expanded = true,
            ManagesFolders = true,
            Folders =
            [
                Folder("work", "Work"),
                Folder("pending-folder:3", "Travel", pending: true, editable: false, acceptsFolders: false),
            ],
        };
        SidebarFixture.Sync(target, [account]);

        var accountRow = target.Single(i => i.AccountId == Account);
        Assert.True(accountRow.AcceptsFolders, "the account row carries the account's capability");
        var work = accountRow.Children.Single(i => i.Tag == "work");
        Assert.True(work.Editable && work.AcceptsFolders && work.AcceptsMessages);
        Assert.False(work.IsPending);
        Assert.Null(work.PendingTip);
        Assert.Equal(1.0, work.RowOpacity);
        var travel = accountRow.Children.Single(i => i.Tag == "pending-folder:3");
        Assert.True(travel.IsPending);
        Assert.Equal("Waiting to sync", travel.PendingLabel);
        Assert.Equal("Waiting to sync", travel.PendingTip);
        Assert.True(travel.RowOpacity < 1.0, "a pending row draws dimmed");
    }
}

File: clients/linux/src/ui/folder_actions_tests.rs
//! What a pane row offers, where a drag may land, and what the dialogs say
//! (`docs/folder-pane.md`, rules 22 to 28).

use mailcal_bindings::{FolderNameCheck, FolderRole, FolderRow, SelectedRow};

use super::{
    Dragged, DropSpot, FolderMenuItem, accepts_drop, account_menu, ancestors, delete_copy,
    folder_menu, move_candidates, name_problem, subtree,
};
use crate::l10n;

fn row(key: &str, name: &str, parent: Option<&str>) -> FolderRow {
    FolderRow {
        key: key.to_owned(),
        name: name.to_owned(),
        role: None,
        unread: 0,
        parent: parent.map(str::to_owned),
        depth: 0,
        has_children: false,
        expanded: false,
        visible: true,
        pending: false,
        in_trash: false,
        editable: true,
        accepts_folders: true,
        accepts_messages: true,
    }
}

/// Inbox (a role folder), `Work` holding `2024` holding `Q1`, and `Home`.
fn tree() -> Vec<FolderRow> {
    let mut inbox = row("inbox", "INBOX", None);
    inbox.role = Some(FolderRole::Inbox);
    inbox.editable = false;
    vec![
        inbox,
        row("work", "Work", None),
        row("2024", "2024", Some("work")),
        row("q1", "Q1", Some("2024")),
        row("home", "Home", None),
    ]
}

fn folder_drag(key: &str, parent: Option<&str>) -> Dragged {
    Dragged::Folder {
        key: key.to_owned(),
        parent: parent.map(str::to_owned),
    }
}

fn mail() -> Dragged {
    Dragged::Mail(SelectedRow::Message {
        account: "a".to_owned(),
        key: "m1".to_owned(),
    })
}

fn spot(folders: &[FolderRow], key: &str) -> DropSpot {
    DropSpot::of(folders.iter().find(|f| f.key == key).unwrap(), folders)
}

#[test]
fn a_row_offers_what_its_flags_allow_and_a_pending_row_nothing() {
    let folders = tree();
    assert_eq!(
        folder_menu(&folders[1]),
        vec![
            FolderMenuItem::NewFolder,
            FolderMenuItem::Rename,
            FolderMenuItem::MoveTo,
            FolderMenuItem::Delete
        ]
    );
    assert_eq!(folder_menu(&folders[0]), vec![FolderMenuItem::NewFolder]);
    let mut pending = row("new", "New", None);
    pending.pending = true;
    pending.editable = false;
    pending.accepts_folders = false;
    pending.accepts_messages = false;
    assert!(folder_menu(&pending).is_empty());
    assert_eq!(account_menu(true), vec![FolderMenuItem::NewFolder]);
    assert!(account_menu(false).is_empty());
}

#[test]
fn move_to_lists_top_level_first_and_never_the_folder_or_what_is_inside_it() {
    let folders = tree();
    let candidates = move_candidates(&folders, "work");
    let parents: Vec<Option<&str>> = candidates.iter().map(|c| c.parent.as_deref()).collect();
    assert_eq!(parents, vec![None, Some("inbox"), Some("home")]);
    assert_eq!(candidates[0].label, l10n::folder_move_top_level());
    // A nested folder is named with the folders it sits in.
    let labels: Vec<String> = move_candidates(&folders, "home")
        .into_iter()
        .map(|c| c.label)
        .collect();
    assert!(
        labels.contains(&"Work / 2024 / Q1".to_owned()),
        "{labels:?}"
    );
    assert!(labels.contains(&l10n::folder_inbox().to_owned()));
}

#[test]
fn a_subtree_and_the_chain_above_a_folder_are_read_from_the_rows() {
    let folders = tree();
    let inside = subtree(&folders, "work");
    assert!(inside.contains("work") && inside.contains("2024") && inside.contains("q1"));
    assert!(!inside.contains("home"));
    assert_eq!(
        ancestors(&folders, "q1"),
        vec!["2024".to_owned(), "work".to_owned()]
    );
}

#[test]
fn a_folder_drops_only_somewhere_it_can_go_in_its_own_account() {
    let folders = tree();
    let work = folder_drag("work", None);
    assert!(accepts_drop("a", &spot(&folders, "home"), "a", &work));
    assert!(
        !accepts_drop("a", &spot(&folders, "home"), "b", &work),
        "another account"
    );
    assert!(
        !accepts_drop("a", &spot(&folders, "work"), "a", &work),
        "onto itself"
    );
    assert!(
        !accepts_drop("a", &spot(&folders, "q1"), "a", &work),
        "inside itself"
    );
    let child = folder_drag("2024", Some("work"));
    assert!(
        !accepts_drop("a", &spot(&folders, "work"), "a", &child),
        "where it is"
    );

    let account = DropSpot::Account {
        manages_folders: true,
    };
    assert!(accepts_drop("a", &account, "a", &child), "to the top");
    assert!(
        !accepts_drop("a", &account, "a", &work),
        "already at the top"
    );
    let fixed = DropSpot::Account {
        manages_folders: false,
    };
    assert!(!accepts_drop("a", &fixed, "a", &child));
}

#[test]
fn mail_drops_on_a_folder_that_takes_it_and_never_on_an_account_row() {
    let mut folders = tree();
    assert!(accepts_drop("a", &spot(&folders, "home"), "a", &mail()));
    assert!(!accepts_drop("a", &spot(&folders, "home"), "b", &mail()));
    folders[4].accepts_messages = false;
    assert!(!accepts_drop("a", &spot(&folders, "home"), "a", &mail()));
    let account = DropSpot::Account {
        manages_folders: true,
    };
    assert!(!accepts_drop("a", &account, "a", &mail()));
}

#[test]
fn a_pending_row_takes_no_drop() {
    let mut folders = tree();
    folders[4].pending = true;
    folders[4].accepts_folders = false;
    folders[4].accepts_messages = false;
    assert!(!accepts_drop("a", &spot(&folders, "home"), "a", &mail()));
    assert!(!accepts_drop(
        "a",
        &spot(&folders, "home"),
        "a",
        &folder_drag("work", None)
    ));
}

#[test]
fn the_name_dialog_says_why_a_name_cannot_be_used_but_not_before_one_is_typed() {
    assert_eq!(name_problem(FolderNameCheck::Valid), None);
    assert_eq!(name_problem(FolderNameCheck::Empty), None);
    assert_eq!(
        name_problem(FolderNameCheck::Taken),
        Some(l10n::folder_name_taken())
    );
    assert_eq!(
        name_problem(FolderNameCheck::Separator),
        Some(l10n::folder_name_separator())
    );
    assert_eq!(
        name_problem(FolderNameCheck::Surrounded),
        Some(l10n::folder_name_surrounded())
    );
    assert_eq!(
        name_problem(FolderNameCheck::Control),
        Some(l10n::folder_name_control())
    );
}

#[test]
fn delete_is_a_move_to_trash_outside_it_and_permanent_inside_it() {
    let outside = delete_copy("Work", false);
    assert_eq!(outside.title, l10n::folder_delete_title("Work"));
    assert_eq!(outside.confirm, l10n::action_move_to_trash());
    assert!(!outside.destructive);
    let inside = delete_copy("Work", true);
    assert_eq!(inside.title, l10n::folder_delete_permanent_title("Work"));
    assert_eq!(inside.message, l10n::folder_delete_permanent_message());
    assert_eq!(inside.confirm, l10n::action_delete_permanently());
    assert!(inside.destructive);
}
